### ISSUE TITLE
The process hung up when trying to stop server

### ISSUE THREAD

**Issue by BossZou** [oatpp/oatpp#173]
I create a sub thread to run my server like this:

```cpp
thread_ptr_ = std::make_shared<std::thread>(&WebServer::StartService, this);
```

This is my function to start server:

```cpp
WebServer::StartService() {
    Config& config = Config::GetInstance();
    std::string port;
    Status status;

    status = config.GetServerConfigWebPort(port);
    AppComponent components = AppComponent(std::stoi(port));
    SwaggerComponent swaggerComponent("192.168.1.57", std::stoi(port));

    /* create ApiControllers and add endpoints to router */
    auto router = components.http_router_.getObject();

    auto doc_endpoints = oatpp::swagger::AsyncController::Endpoints::createShared();

    auto user_controller = WebController::createShared();
    user_controller->addEndpointsToRouter(router);

    doc_endpoints->pushBackAll(user_controller->getEndpoints());

    auto swaggerController = oatpp::swagger::AsyncController::createShared(doc_endpoints);
    swaggerController->addEndpointsToRouter(router);

    /* create server */
    server_ptr_ = std::make_unique<oatpp::network::server::Server>(components.server_connection_provider_.getObject(),
                                                                   components.server_connection_handler_.getObject());

    // start synchronously
    server_ptr_->run();

    return Status::OK();
}
```

I try to stop my web server just to call `Stop()` like this:

```cpp
if (server_ptr_ != nullptr) {
        server_ptr_->stop();  // cStop server
    }
```

```cpp
    if (thread_ptr_ != nullptr) {
        thread_ptr_->join();
        thread_ptr_ = nullptr;
    }
```

I join  web thread, and wait it finishing but the process hung up. How I protect it from hanging up?

I use async endpoint, and my server connection provider is an object of `SimpleTCPConnectionProvider`

**Comment by BossZou** [comment]
I think server hang up because it was waiting I/O event. When I call method `Stop`, it just assign `STATUS_STOPPING`  to server status flag `m_status` rather than notify the thread which was waiting I/O event. Am I right ?

**Comment by lganzzzo** [comment]
Hello @BossZou ,

Yes, you are right, the server is waiting for new connections on the blocking `accept()` call.
In order to unblock it call the:

```cpp
server->stop();
connectionProvider->getConnection(); // - it will trigger the blocking accept call.
```
It's a bit inconvenient, but that's how things are for now.

Regards,
Leonid

**Comment by BossZou** [comment]
I see. Does oatpp use Blocking I/O model currently ? Will is support Non_blocking I/O in the future ?

**Comment by lganzzzo** [comment]
For connection accepting oatpp uses blocking `accept`. After connection is accepted its handed to the `ConnectionHandler`.

- If you are using `AsyncHttpConnectionHandler` then connection is switched to the `non-blocking` `IOMode` and futher processing is done by oatpp-coroutines.
So in this case you get full non-blocking IO, and you can serve thousands connection simultaneously regardless of how many threads you have.

- if you are using `HttpConnectionHandler` then you get multithreaded blocking IO having one thread per connection.

So oatpp supports both blocking and non-blocking IO. Connection accepting is done in blocking manner in separate thread.

Hope this has clarified things a little bit.
Please let me know if you have more questions.

Regards,
Leonid

**Comment by BossZou** [comment]
> Hello @BossZou ,
> 
> Yes, you are right, the server is waiting for new connections on the blocking `accept()` call.
> In order to unblock it call the:
> 
> ```c++
> server->stop();
> connectionProvider->getConnection(); // - it will trigger the blocking accept call.
> ```
> 
> It's a bit inconvenient, but that's how things are for now.
> 
> Regards,
> Leonid

Which type of `connectionProvider` ? In my case I call `getConnection()` of an object of 
 class `SimpleTCPConnectionProvider`. However, the method `getConnection()` blocked. I think I'm wrong. So how to slove it ?

**Comment by lganzzzo** [comment]
Sorry, my bad, didn't clarified that.

In order to unblock `ServerConnectionProvider`, you have to create `ClientConnectionProvider` which will try to connect to the server port and unblock the accept.

So create client connection provider, and call `getConnection()` on it.

---

This is ugly workaround but it works. Proper server stop is not implemented in oatpp yet.

**Comment by BossZou** [comment]
All right. There is any plan to solve it ?

**Comment by lganzzzo** [comment]
It's a low priority currently, but it will be solved sooner or later.
You can expect it to be resolved somewhere in March-May.

**Comment by BossZou** [comment]
I create `ClientConnectionProvider` and used it when I tried to close my server. 
I use create it like this:
```
OATPP_CREATE_COMPONENT(std::shared_ptr<oatpp::network::ClientConnectionProvider>, client_connection_provider_)([this] {
        return std::static_pointer_cast<oatpp::network::ClientConnectionProvider>(
            oatpp::network::client::SimpleTCPConnectionProvider::createShared("127.0.0.1", this->port_)
        );
    }());
```
Then I close my server like this:
```
if (server_ptr_ != nullptr) {
        server_ptr_->stop();

        OATPP_COMPONENT(std::shared_ptr<oatpp::network::ClientConnectionProvider>, client_connection_provider);
        client_connection_provider->getConnection();
//        requestExecutor->getConnection();
        //
    }
```

However. An error occurred.
```
terminate called after throwing an instance of 'std::runtime_error'
  what():  [oatpp::base::memory::MemoryPool::freeByEntryHeader()]: Invalid EntryHeader
```

Can you help me?

**Comment by BossZou** [comment]
I track the stack and screenshot as follow:
![stackmap_bottom](https://user-images.githubusercontent.com/40255591/71732155-81b4f380-2e81-11ea-85f7-dc2d1a926a15.png)
![stackmap_top](https://user-images.githubusercontent.com/40255591/71732163-87123e00-2e81-11ea-8f2c-95f043c0021f.png)

It seam as a error occurred when deconstructing. I have no idea to deal with it.

**Comment by lganzzzo** [comment]
Hey @BossZou ,

It is not so easy to advise on this one.
It seems like you have a memory leak somewhere in your code.

According to stack-trace my best assumption is that you are stopping the server, the program exits the block in which your AppComponents were created and destroy all the components. 
However, it might be the case that coroutines spawned by `async::Executor` are still alive and trying to access destroyed resources (the router).

So the solution, in this case, is to wait until the `async::Executor` is finished and exit block afterward.

Pseudocode:
```
{ // block

  AppComponent components; // create components.

  Server server(...);
  
  server.run();

  OATPP_COMPONENT(std::shared_ptr<oatpp::async::Executor>, executor); // get executor
  executor->waitTasksFinished(); // wait all the coroutines;
  executor->join(); // joint executor threads.

}
```

You can see how it's done in the tests [FullAsyncTest](https://github.com/oatpp/oatpp/blob/master/test/oatpp/web/FullAsyncTest.cpp#L282)

Please let me know if it helped.

Regards,
Leonid

**Comment by BossZou** [comment]
It hung up when executed `executor->join();`

**Comment by lganzzzo** [comment]
I see. You should also call `close` on your `ConnectionHandler` - which is AsyncHttpConnectionHandler

**Comment by BossZou** [comment]
Where should I call `stop` on your `AsyncHttpConnectionHandler` ? I add it like this:
```
server_ptr_->run();

    OATPP_COMPONENT(std::shared_ptr<oatpp::network::server::ConnectionHandler>, server_connection_handler);
    server_connection_handler->stop();

    OATPP_COMPONENT(std::shared_ptr<oatpp::async::Executor>, executor);
    executor->waitTasksFinished();
    executor->join();
```

But it  still throw Exception:
```
terminate called after throwing an instance of 'std::runtime_error'
  what():  [oatpp::base::memory::MemoryPool::freeByEntryHeader()]: Invalid EntryHeader
```

**Comment by lganzzzo** [comment]
Hey @BossZou ,

I have created the full-working example for you from oatpp-starter-async.
It waits 10s then stops the server.

This is my App.cpp

```cpp

#include "./controller/MyController.hpp"
#include "./AppComponent.hpp"

#include "oatpp/network/server/Server.hpp"

#include <iostream>
#include <thread>

void run() {

  /* Register Components in scope of run() method */
  AppComponent components;

  /* Get router component */
  OATPP_COMPONENT(std::shared_ptr<oatpp::web::server::HttpRouter>, router);

  /* Create MyController and add all of its endpoints to router */
  auto myController = std::make_shared<MyController>();
  myController->addEndpointsToRouter(router);

  /* Get connection handler component */
  OATPP_COMPONENT(std::shared_ptr<oatpp::network::server::ConnectionHandler>, connectionHandler);

  /* Get connection provider component */
  OATPP_COMPONENT(std::shared_ptr<oatpp::network::ServerConnectionProvider>, connectionProvider);

  /* Create server which takes provided TCP connections and passes them to HTTP connection handler */
  oatpp::network::server::Server server(connectionProvider, connectionHandler);

  /* Priny info about server port */
  OATPP_LOGI("MyApp", "Server running on port %s", connectionProvider->getProperty("port").getData());

  // Start timer thread
  std::thread timer([connectionHandler, &server]{

    OATPP_LOGD("MyApp", "Waiting 10s");

    std::this_thread::sleep_for(std::chrono::seconds(10));

    OATPP_LOGD("MyApp", "Stopping server...");

    server.stop();

    OATPP_COMPONENT(std::shared_ptr<oatpp::network::ClientConnectionProvider>, connectionProvider);
    connectionProvider->getConnection();

  });

  /* Run server */
  server.run();

  connectionHandler->stop();

  OATPP_COMPONENT(std::shared_ptr<oatpp::async::Executor>, executor); // get executor
  executor->waitTasksFinished(); // wait all the coroutines;
  executor->join(); // joint executor threads.

  timer.join();

}

int main(int argc, const char * argv[]) {

  oatpp::base::Environment::init();

  run();

  /* Print how much objects were created during app running, and what have left-probably leaked */
  /* Disable object counting for release builds using '-D OATPP_DISABLE_ENV_OBJECT_COUNTERS' flag for better performance */
  std::cout << "\nEnvironment:\n";
  std::cout << "objectsCount = " << oatpp::base::Environment::getObjectsCount() << "\n";
  std::cout << "objectsCreated = " << oatpp::base::Environment::getObjectsCreated() << "\n\n";

  oatpp::base::Environment::destroy();

  return 0;
}

```

And this is my AppComponent.hpp

```cpp

#ifndef AppComponent_hpp
#define AppComponent_hpp

#include "oatpp/web/server/AsyncHttpConnectionHandler.hpp"
#include "oatpp/web/server/HttpRouter.hpp"

#include "oatpp/network/client/SimpleTCPConnectionProvider.hpp"
#include "oatpp/network/server/SimpleTCPConnectionProvider.hpp"

#include "oatpp/parser/json/mapping/ObjectMapper.hpp"

#include "oatpp/core/macro/component.hpp"

/**
 *  Class which creates and holds Application components and registers components in oatpp::base::Environment
 *  Order of components initialization is from top to bottom
 */
class AppComponent {
public:

  /**
   * Create Async Executor
   */
  OATPP_CREATE_COMPONENT(std::shared_ptr<oatpp::async::Executor>, executor)([] {
    return std::make_shared<oatpp::async::Executor>(
      4 /* Data-Processing threads */,
      1 /* I/O threads */,
      1 /* Timer threads */
    );
  }());
  
  /**
   *  Create ConnectionProvider component which listens on the port
   */
  OATPP_CREATE_COMPONENT(std::shared_ptr<oatpp::network::ServerConnectionProvider>, serverConnectionProvider)([] {
    return oatpp::network::server::SimpleTCPConnectionProvider::createShared(8000);
  }());

  OATPP_CREATE_COMPONENT(std::shared_ptr<oatpp::network::ClientConnectionProvider>, clientConnectionProvider)([] {
    return oatpp::network::client::SimpleTCPConnectionProvider::createShared("localhost", 8000);
  }());
  
  /**
   *  Create Router component
   */
  OATPP_CREATE_COMPONENT(std::shared_ptr<oatpp::web::server::HttpRouter>, httpRouter)([] {
    return oatpp::web::server::HttpRouter::createShared();
  }());
  
  /**
   *  Create ConnectionHandler component which uses Router component to route requests
   */
  OATPP_CREATE_COMPONENT(std::shared_ptr<oatpp::network::server::ConnectionHandler>, serverConnectionHandler)([] {
    OATPP_COMPONENT(std::shared_ptr<oatpp::web::server::HttpRouter>, router); // get Router component
    OATPP_COMPONENT(std::shared_ptr<oatpp::async::Executor>, executor); // get Async executor component
    return oatpp::web::server::AsyncHttpConnectionHandler::createShared(router, executor);
  }());
  
  /**
   *  Create ObjectMapper component to serialize/deserialize DTOs in Contoller's API
   */
  OATPP_CREATE_COMPONENT(std::shared_ptr<oatpp::data::mapping::ObjectMapper>, apiObjectMapper)([] {
    return oatpp::parser::json::mapping::ObjectMapper::createShared();
  }());

};

#endif /* AppComponent_hpp */

```

Regards,
Leonid

**Comment by lganzzzo** [comment]
Closing this one...
@BossZou , please feel free to reopen this ticket, in case some questions left on this matter.

**Comment by darren-harton** [comment]
Update for future readers: The above example no longer works unless `executor->stop();` is called before `executor->waitTasksFinished()`.

**Comment by lganzzzo** [comment]
Hello @darren-harton ,

Thanks for the update!

That is because in previous versions `AsyncConnectionHandler` was calling `stop()` on the executor.
This changes in https://github.com/oatpp/oatpp/commit/121d079e7d8a2d77195a21ec5ecdd1f7ed4fc325#diff-f79aea0b01e2432d1def2587a6ad3a07 .

However, please note, that the correct stoppage should be `executor->stop();` after `executor->waitTasksFinished();`

```cpp

  ...

  /* Run server */
  server.run();

  /* Server was stopped. Now stopping the AsyncExecutor */
  executor->waitTasksFinished(); 
  executor->stop(); // <--- stop after tasks finished!
  executor->join(); 

  ...

```

Please let me know if it also works for you.

Regards,
Leonid
